src/hotspot/share/runtime/stackWatermark.cpp

```diff
@@ -0,0 +1,301 @@
+    }
+  }
+}
+
+StackWatermarkFramesIterator::StackWatermarkFramesIterator(StackWatermark& owner) :
+    _jt(owner._jt),
+    _caller(0),
+    _callee(0),
+    _frame_stream(owner._jt, true /* update_registers */, false /* process_frames */),
+    _owner(owner),
+    _is_done(_frame_stream.is_done()) {
+}
+
+frame& StackWatermarkFramesIterator::current() {
+  return *_frame_stream.current();
+}
+
+RegisterMap& StackWatermarkFramesIterator::register_map() {
+  return *_frame_stream.register_map();
+}
+
+bool StackWatermarkFramesIterator::has_next() const {
+  return !_is_done;
+}
+
+void StackWatermarkFramesIterator::next() {
+  _frame_stream.next();
+  _is_done = _frame_stream.is_done();
+}
+
+StackWatermark::StackWatermark(JavaThread* jt, StackWatermarkKind kind, uint32_t epoch) :
+    _state(StackWatermarkState::create(epoch, true /* is_done */)),
+    _watermark(0),
+    _next(NULL),
+    _jt(jt),
+    _iterator(NULL),
+    _lock(Mutex::tty - 1, "stack_watermark_lock", true, Mutex::_safepoint_check_never),
+    _kind(kind) {
+}
+
+StackWatermark::~StackWatermark() {
+  delete _iterator;
+}
+
+#ifdef ASSERT
+void StackWatermark::assert_is_frame_safe(const frame& f) {
+  MutexLocker ml(&_lock, Mutex::_no_safepoint_check_flag);
+  assert(is_frame_safe(f), "Frame must be safe");
+}
+#endif
+
+// A frame is "safe" if it *and* its caller have been processed. This is the invariant
+// that allows exposing a frame, and for that frame to directly access its caller frame
+// without going through any hooks.
+bool StackWatermark::is_frame_safe(const frame& f) {
+  assert(_lock.owned_by_self(), "Must be locked");
+  uint32_t state = Atomic::load(&_state);
+  if (!processing_started(state)) {
+    return false;
+  }
+  if (processing_completed(state)) {
+    return true;
+  }
+  return reinterpret_cast<uintptr_t>(f.sp()) < _iterator->caller();
+}
+
+void StackWatermark::start_processing_impl(void* context) {
+  log_info(stackbarrier)("Starting stack processing for tid %d",
+                         _jt->osthread()->thread_id());
+  delete _iterator;
+  if (_jt->has_last_Java_frame()) {
+    _iterator = new StackWatermarkFramesIterator(*this);
+    // Always process three frames when starting an iteration.
+    //
+    // The three frames corresponds to:
+    // 1) The callee frame
+    // 2) The caller frame
+    // This allows a callee to always be able to read state from its caller
+    // without needing any special barriers.
+    //
+    // 3) An extra frame to deal with unwinding safepointing on the way out.
+    // Sometimes, we also call into the runtime to on_unwind(), but then
+    // hit a safepoint poll on the way out from the runtime.
+    _iterator->process_one(context);
+    _iterator->process_one(context);
+    _iterator->process_one(context);
+  } else {
+    _iterator = NULL;
+  }
+  update_watermark();
+}
+
+void StackWatermark::yield_processing() {
+  update_watermark();
+  MutexUnlocker mul(&_lock, Mutex::_no_safepoint_check_flag);
```

Comment:
This seems a little dubious - is it just a heuristic? There is no guarantee that unlocking the Mutex will allow another thread to claim it before this thread re-locks it.

Reply:
It is indeed just a heuristic. There is no need for a guarantee.